rmc/src/service.rs

```diff
@@ -53,310 +43,250 @@ impl<H, MK, SCH> Service<H, MK, SCH>
+        Service { scheduler, handler }
     }
 
-    fn handle_message(
-        &mut self,
-        message: RmcIncomingMessage<H, MK::Signature, MK::PartialMultisignature>,
-    ) {
-        match message {
-            RmcIncomingMessage::StartRmc(hash) => {
-                debug!(target: LOG_TARGET, "starting rmc for {:?}", hash);
-                let signed_hash = self.handler.on_start_rmc(hash);
+    /// Signs the given `hash` and adds the signature to the collection. If the given `hash`
+    /// completes the multisignature, it is scheduled for the broadcasts and then returned.
+    /// If the multisignature is not completed, `None` is returned. If the multisignature was
+    /// already completed when starting rmc, no tasks are scheduled. Otherwise the signed hash
+    /// is scheduled for the broadcasts.
+    pub fn start_rmc(&mut self, hash: H) -> Option<Multisigned<H, MK>> {
+        debug!(target: LOG_TARGET, "starting rmc for {:?}", hash);
+        match self.handler.on_start_rmc(hash) {
+            OnStartRmcResponse::SignedHash(signed_hash) => {
                 self.scheduler
-                    .add_task(RmcHash::SignedHash(signed_hash.into_unchecked()));
+                    .add_task(Message::SignedHash(signed_hash.into_unchecked()));
+            }
+            OnStartRmcResponse::MultisignedHash(multisigned) => {
+                self.scheduler.add_task(Message::MultisignedHash(
+                    multisigned.clone().into_unchecked(),
+                ));
+                return Some(multisigned);
             }
-            RmcIncomingMessage::RmcHash(RmcHash::MultisignedHash(unchecked)) => {
+            OnStartRmcResponse::Noop => {}
+        }
+        None
+    }
+
+    /// Processes a message which can be of two types. If the message is a hash signed by one
+    /// person, it adds it to the collective signature. If it completes the multisignature, it is
+    /// returned. Otherwise `None` is returned. If the message is a multisigned hash, it returns
+    /// the multisignature, if we haven't seen it before. Otherwise `None` is returned.
+    pub fn process_message(
+        &mut self,
+        message: Message<H, MK::Signature, MK::PartialMultisignature>,
+    ) -> Option<Multisigned<H, MK>> {
+        match message {
+            Message::SignedHash(unchecked) => match self.handler.on_signed_hash(unchecked) {
+                Ok(Some(multisigned)) => {
+                    self.scheduler.add_task(Message::MultisignedHash(
+                        multisigned.clone().into_unchecked(),
+                    ));
+                    return Some(multisigned);
+                }
+                Ok(None) => {}
+                Err(error) => {
+                    warn!(target: LOG_TARGET, "failed handling multisigned hash: {}", error);
+                }
+            },
+            Message::MultisignedHash(unchecked) => {
                 match self.handler.on_multisigned_hash(unchecked.clone()) {
```

Comment:
You don't need to clone here? Later you can call `multisigned.clone().into_unchecked()` or provide `.as_unchecked` for `Multisigned`...